Marc21Record Validation

bengaal opened the pull request:
This PR adds validation of the printed `Marc21Record` with an [external schema](https://www.loc.gov/standards/marcxml/schema/MARC21slim.xsd) defined by the Library of Congress.

bengaal committed:
bugfix(Marc21Record): replace empty string with space as default value

Upon validation it became clear, that the default value for ind1 and
ind2 should be space instead of empty string.

bengaal committed:
bugfix(Marc21Record): add leader to method call

bengaal committed:
bugfix(Marc21Record): fix indentations

Indentations were incorrect.

bengaal committed:
bugfix(Marc21Record): remove erroneous comma from xml element

bengaal committed:
bugfix(Marc21Record): add closing tag for subfield

Subfield was missing a closing tag.

bengaal committed:
feature(Marc21Record): add validation with schema

bengaal committed:
bugfix(Marc21Record): replace conflict causing parts

A Marc21Record must have a leader. In order to comply with the schema
a faked leader is added.
Xml version, schema references are added to the record.

utnapischtim commented on tests/conftest.py:
```diff
@@ -87,3 +87,813 @@ def teardown():
+@pytest.fixture(scope="function")
+def pure_record_merged():
+    """A Pure record with all possible attributes."""
+    return {
```
would it be possible to create such big json objects in a separated '*.json' file and load that file?

bengaal replied:
Take a look at https://github.com/tu-graz-library/invenio-rdm-pure/pull/33/commits/4909820b9e0e25d77f1c697b3a90252564a593b4. This will be merged in https://github.com/tu-graz-library/invenio-rdm-pure/pull/32.

utnapischtim replied:
yes, i saw it afterwards. could you please then through out those unnecessary commits

bengaal replied:
As soon as https://github.com/tu-graz-library/invenio-rdm-pure/pull/32 is merged, this branch will be rebased and the commits will be gone. Without those commits the tests would fail, blocking the merge.

bengaal replied:
I have removed the commits of https://github.com/tu-graz-library/invenio-rdm-pure/pull/32.

utnapischtim commented on invenio_rdm_pure/source/rdm/marc21_record.py:
```diff
@@ -46,15 +46,20 @@ def __init__(self, code: str = "", value: str = ""):
 
-    def __init__(self, leader: str = ""):
+    def __init__(
+        self, leader: str = "97628dxefp 201109bOq4500"
```
what are those magic numbers?

bengaal replied:
The schema doesn't allow an empty leader, so I added a value, which conforms with the schema. Could you please give me more information how the leader should be handled in imports?

bengaal replied:
The default value has been moved to a constant value.

bengaal committed:
refactor(Marc21Record): move xml tag conversion methods to proper classes

bengaal committed:
refactor(Marc21Record): move leader placeholder to a constant variable

bengaal committed:
refactor(Marc21Record): rename pretty_print method

bengaal committed:
refactor(Marc21Record): restructure leader to xml tag method

bengaal committed:
refactor(Marc21Record): restructure validation method

bengaal committed:
refactor(Marc21Record): restructure string building

I found evidence that suggests there is no significant speed gain
with the previous way of concatenation.
https://waymoot.org/home/python_string
Concatenating the string this way has the benefit of improved
readability.

utnapischtim commented on invenio_rdm_pure/source/rdm/marc21_record.py:
```diff
@@ -142,3 +151,14 @@ def add_unique_value(
             self.datafields.append(datafield)
+
+    @staticmethod
+    def validate_marc21_xml_string(record: str) -> str:
```
Use this method to test the Converter() output.

bengaal replied:
Please see https://github.com/tu-graz-library/invenio-rdm-pure/pull/33/commits/f4c46027ab75a52fdf9ecd9d15d84bbd60aeb754, I hope this is what you meant.

utnapischtim commented on invenio_rdm_pure/source/rdm/marc21_record.py:
```diff
@@ -38,68 +65,50 @@ def __init__(self, code: str = "", value: str = ""):
 
-    def __init__(self, leader: str = ""):
+    LEADER_PLACEHOLDER = (
+        "97628dxefp 201109bOq4500"  # TODO: find a way to generate proper leaders
```
https://www.loc.gov/marc/bibliographic/bdleader.html

use: 00000nam#a2200000zca4500

bengaal committed:
refactor(leader): use suggested leader

bengaal committed:
build(MANIFEST.in): add .xsd schema to manifest

bengaal committed:
modification(Marc21Record): change validation method use

bengaal committed:
build(dependency): bump dependency version

Bump dpeendency version in order to be compatible with
invenio-cli 0.21.0
